### PR TITLE
Constructing and serializing solana transactions

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/20697

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/solana_utils.cc [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/solana_utils.h"
+
+#include "base/check.h"
+#include "brave/third_party/bitcoin-core/src/src/base58.h"
+
+namespace brave_wallet {
+
+void CompactU16Encode(uint16_t u16, std::vector<uint8_t>* compact_u16) {
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_utils.cc:13]: As mentioned in the header, doc is at https://docs.solana.com/developing/programming-model/transactions#compact-u16-format.
Ref impl: https://github.com/solana-labs/solana-web3.js/blob/master/src/util/shortvec-encoding.ts#L15

File: components/brave_wallet/browser/solana_utils_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#include <map>
+#include <vector>
+
+#include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace brave_wallet {
+
+TEST(SolanaUtilsUnitTest, CompactU16Encode) {
+  std::map<uint16_t, std::vector<uint8_t>> input_output_map = {
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_utils_unittest.cc:17]: Test cases from https://docs.rs/solana-program/1.7.11/src/solana_program/short_vec.rs.html#287

File: components/brave_wallet/browser/solana_utils.cc [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  DCHECK(compact_u16);
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/solana_utils.cc:14]: Fixed in https://github.com/brave/brave-core/pull/12120/commits/572b40cd7f89764f35430dd8597f832eb4cc511c

File: components/brave_wallet/browser/solana_utils.cc [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      elem |= 0x80;
+      compact_u16->push_back(elem);
+    }
+  }
+}
+
+bool Base58Decode(const std::string& str,
+                  std::vector<uint8_t>* ret,
+                  int max_ret_len) {
+  return DecodeBase58(str, *ret, max_ret_len);
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/solana_utils.cc:34]: Used a DCHECK instead and added a test in https://github.com/brave/brave-core/pull/12120/commits/572b40cd7f89764f35430dd8597f832eb4cc511c.

File: components/brave_wallet/browser/solana_instruction.cc [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+SolanaInstruction::~SolanaInstruction() = default;
+
+// Each instruction specifies a single program, a subset of the transaction's
+// accounts that should be passed to the program, and a data byte array that is
+// passed to the program. See
+// https://docs.solana.com/developing/programming-model/transactions#instructions
+// for more details.
+bool SolanaInstruction::Serialize(
+    const std::vector<SolanaAccountMeta>& account_metas,
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/solana_instruction.cc]: https://github.com/brave/brave-core/pull/12120/commits/138ba3e385103a40aa55e5d774cf4ba80a37b247
Hope this helps with the readability, let me know if it's still hard to read.

File: components/brave_wallet/browser/solana_message.cc [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  DCHECK(unique_account_metas);
+  unique_account_metas->clear();
+
+  std::vector<SolanaAccountMeta> account_metas;
+  // Get accounts from each instruction.
+  for (const auto& instruction : instructions_) {
+    account_metas.insert(account_metas.end(), instruction.GetAccounts().begin(),
+                         instruction.GetAccounts().end());
+  }
+  // Get program ID from each instruction.
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/solana_message.cc]: Yeah, no need, updated in https://github.com/brave/brave-core/pull/12120/commits/c7dd900db406b75cb3e56e0dd23db0e5a52ce59a and also switched to stable_sort to ensure the order of equal elements will be preserved.

**Review by darkdh — APPROVED** [review]

**Pull request merged**